Add Riemann sink

jfoy opened the pull request:
Fixes #293

jfoy commented:
CCLA signed under Whitepages.

jfoy commented:
/cc @aphyr

vishh commented:
cc @vmarmol

jfoy commented:
Seems I can't see the failing e2e test; is there a way to retrieve the result?

vmarmol commented:
We're having issues with the e2e, I'd ignore it for now.

jfoy commented:
OK. What else do we need to review before this can be merged?

vishh commented on deploy/kube-config/riemann/heapster-service.json:
```diff
@@ -0,0 +1,22 @@
+{
```
Can we place a symlink to an existing service file instead of creating a new one?

jfoy replied:
Done.

vishh commented on riemann/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM debian:8.1
```
Does this have to be maintained here or can we reuse an existing reimann docker image?

jfoy replied:
We'll need to build our own image even if we derive it from an existing base, because we still need to supply an appropriate `/etc/riemann/riemann.config`.

vishh replied:
Got it! Thanks for explaining!

vishh commented on sinks/riemann/driver.go:
```diff
@@ -0,0 +1,211 @@
+		// Source: A short, machine-understantable string that gives the
+		// component reporting this event
+		Host:       event.Source.Host,
+		Attributes: event.Labels, // + event.Annotations,
```
Is annotations a TODO here?

jfoy replied:
Yes, done.

vishh commented:
Overall LGTM. Sorry for the delay in reviewing this PR. This needs a rebase as well.

vishh commented:
And do squash your commits, once you have gone through the comments.

jfoy commented:
Great -- I'll get this caught up.

vishh commented:
@jfoy did you get a chance to work on this PR?

jfoy commented:
I will -- thanks for your patience.

vishh commented:
@jfoy: Ping.

jfoy commented:
Not forgotten. We've been iterating on this internally, and I have a set of changes ready to go, which I'll push with the rebase and squash.

jfoy commented:
Done and squashed.

jfoy commented:
What's going on with our CCLA? We signed one from Whitepages, which seems to have vanished.

jfoy commented:
Fixing dependencies to unblock unit tests. (How embarrassing.)

vishh commented on Makefile:
```diff
@@ -2,10 +2,13 @@ all: build
 
 deps:
 	go get github.com/tools/godep
+	go get github.com/progrium/go-extpoints
```
cc @mvdan

mvdan replied:
This should not be here since we use the vendored go-extpoints version as seen in https://github.com/kubernetes/heapster/blob/8e4a8ba63e75fe234bedb4171b35553d6ea18981/Godeps/Godeps.json#L6

jfoy replied:
Done.

vishh commented on deploy/kube-config/riemann/heapster-controller.json:
```diff
@@ -0,0 +1,36 @@
+{
```
Can we avoid adding a separate controller? I'm trying to get rid of the existing ones and document using the remote API for enabling sinks. For now, just documenting the reimann sink should work I think.

jfoy replied:
Done. I'll add corresponding documentation.

vishh commented on deploy/kube-config/riemann/heapster-service.json:
```diff
@@ -0,0 +1 @@
+../influxdb/heapster-service.json
```
Same as above

jfoy replied:
Done.

vishh commented on deploy/kube-config/riemann/riemann-service.json:
```diff
@@ -0,0 +1,36 @@
+    },
+    "name": "riemann-heapster"
+  },
+  "spec": {
```
Does reimann have a web UI that needs to be exposes outside of the cluster?

jfoy replied:
Not in the engine itself. There's a separately-distributed [dashboard](https://github.com/aphyr/riemann-dash), which in practice I would package as a separate container. It wouldn't be crazy for it to be in the same pod, though.

vishh replied:
Ok. Shall we ship the UI by default in the same Pod? It will definitely help users. 
Also, can we route data to InfluxDB via Reimann or is it recommended to directly store in InfluxDB (or some other timeseries store)?

jfoy replied:
We can route to InfluxDB via Riemann, and you probably typically would. It's our intent to do so. Setting up that configuration is a larger task than I wanted to take on in this PR.

vishh replied:
Agreed. Any thoughts on the UI?

jfoy replied:
It's a good idea, but I'm tempted to defer. We have run with riemann-dash in the past, and we're not currently doing so; it's a useful tool for exploration and certain visualizations. In practice, we stream our metrics from Riemann to Grafana dashboards.

vishh replied:
Ah! Ok. Thanks for explaining. In that case just some pointers to existing
documentation should do.

On Tue, Sep 22, 2015 at 4:00 PM, Jack Foy <notifications@github.com> wrote:

> In deploy/kube-config/riemann/riemann-service.json
> <https://github.com/kubernetes/heapster/pull/379#discussion_r40155265>:
>
> > @@ -0,0 +1,36 @@
> > +{
> > +  "apiVersion": "v1beta3",
> > +  "kind": "Service",
> > +  "metadata": {
> > +    "labels": {
> > +      "kubernetes.io/cluster-service": "true",
> > +      "name": "riemann-heapster"
> > +    },
> > +    "name": "riemann-heapster"
> > +  },
> > +  "spec": {
>
> It's a good idea, but I'm tempted to defer. We have run with riemann-dash
> in the past, and we're not currently doing so; it's a useful tool for
> exploration and certain visualizations. In practice, we stream our metrics
> from Riemann to Grafana dashboards.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/379/files#r40155265>.
>

vishh commented on sinks/riemann/driver.go:
```diff
@@ -0,0 +1,230 @@
+	return []sink_api.ExternalSink{sink}, nil
+}
+
+func New(getClient func() riemannClient, conf riemannConfig) (sink_api.ExternalSink, error) {
```
Can this be private?

jfoy replied:
Done.

vishh commented:
@jfoy: This is amazing! Can't wait to try it out. Can you post some documentation on how users can use Reimann with heapster? Maybe some sample alert configs.

Other than documentation, everything else LGTM.

mvdan commented on .travis.yml:
```diff
@@ -13,6 +13,8 @@ install:
  - go get golang.org/x/tools/cmd/cover
  - go get github.com/mattn/goveralls
+ - go get github.com/onsi/ginkgo
+ - go get github.com/onsi/gomega
```
These should probably be vendored. If not, I'm thinking that the riemann tests may break at any point.

Didnt `godep save ./...` add them?

jfoy replied:
No, my mistake. I hadn't run `godeps save`.

jfoy commented:
That was weird. Reverting to the prior state until I figure out what's going on with `godep`.

mvdan commented:
@jfoy I can take care of the godep stuff if you like - remove all the godep changes and I'll do a PR on top of yours.

jfoy commented:
@mvdan Awesome, thank you. I would like to understand where the variation comes from, though -- if I do a 'godep restore' followed by a 'godep save -t', why does the resulting Godeps/Godeps.json show so many deleted packages?

mvdan commented:
Running exactly those commands, the answer is simple - you have to give `./...` as an argument to `godep save`. If you don't, it will assume `.`, thus saving only the dependencies for the main heapster package.

jfoy commented:
Here's what I see:
```
201345m0:heapster jfoy$ godep restore
201345m0:heapster jfoy$ git status
On branch riemann_sink
Your branch is up-to-date with 'jfoy/riemann_sink'.
nothing to commit, working directory clean
201345m0:heapster jfoy$ godep save -t ./... github.com/progrium/go-extpoints
201345m0:heapster jfoy$ git status
On branch riemann_sink
Your branch is up-to-date with 'jfoy/riemann_sink'.
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   Godeps/Godeps.json
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/LICENSE
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/dce.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/doc.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/hash.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/node.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/time.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/util.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/uuid.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/uuid_test.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/version1.go
	deleted:    Godeps/_workspace/src/code.google.com/p/go-uuid/uuid/version4.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/flags.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/flags_freebsd.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/flags_linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/flags_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mount.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mount_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mounter_freebsd.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mounter_linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mounter_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mountinfo.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mountinfo_freebsd.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mountinfo_linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mountinfo_linux_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/mountinfo_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/sharedsubtree_linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/mount/sharedsubtree_linux_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/units/duration.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/units/duration_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/units/size.go
	deleted:    Godeps/_workspace/src/github.com/docker/docker/pkg/units/size_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/cgroups.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/cgroups_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/cgroups_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/apply_raw.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/blkio.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/blkio_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/cpu.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/cpu_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/cpuacct.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/cpuset.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/cpuset_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/devices.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/devices_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/freezer.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/freezer_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/fs_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/hugetlb.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/hugetlb_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/memory.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/memory_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/net_cls.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/net_cls_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/net_prio.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/net_prio_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/perf_event.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/stats_util_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/util_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/utils.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/fs/utils_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/stats.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/systemd/apply_nosystemd.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/systemd/apply_systemd.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/cgroups/utils.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/cgroup.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/config.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/config_test.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/config_unix.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/device.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/device_defaults.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/hugepage_limit.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/interface_priority_map.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/mount.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/namespaces.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/namespaces_syscall.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/namespaces_syscall_unsupported.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/namespaces_unix.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/namespaces_windows.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/network.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/configs/validate/config.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/proc.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/setns_linux.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/syscall_linux_386.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/syscall_linux_64.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/syscall_linux_arm.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/sysconfig.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/sysconfig_notcgo.go
	deleted:    Godeps/_workspace/src/github.com/docker/libcontainer/system/xattrs_linux.go
	modified:   Godeps/_workspace/src/github.com/prometheus/client_golang/text/bench_test.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/apis/experimental/install/install.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/helpers.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/latest/latest.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/register.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/types.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/v1/conversion.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/v1/register.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api/v1/types.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/auth_loaders.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/client_config.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/doc.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/loader.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/merged_client_builder.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/overrides.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/validation.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/probe/doc.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/probe/exec/exec.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/probe/http/http.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/probe/probe.go
	deleted:    Godeps/_workspace/src/k8s.io/kubernetes/pkg/probe/tcp/tcp.go

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Godeps/_workspace/src/github.com/onsi/

no changes added to commit (use "git add" and/or "git commit -a")
201345m0:heapster jfoy$
```

jfoy commented:
@vishh I'm pulling together some config samples.

mvdan commented:
@jfoy funnily enough, I get the same thing you get when running save on master. It obviously breaks:

```
sources/kube_factory.go:30:2: cannot find package "k8s.io/kubernetes/pkg/client/unversioned/clientcmd" in any of:
```

Some of the things it does, like removing `go-uuid`, are fine - but others aren't. The best I can come up with now is that godep must be broken. We might be doing something wrong here, but the problem is just too obvious.

vishh commented:
@cadvisorJenkinsBot: test this please

jfoy commented:
I can't tell by looking at the output -- is this another artifact of the e2e pipeline?

mvdan commented:
@jfoy see #601 - could you rebase your PR and try his trick of removing Godeps before running save?

jimmidyson commented:
It may be enough to just rebase & do `godep save ./... github.com/progrium/go-extpoints` now that godeps are cleaned up a bit. I'd try that first.

mvdan commented:
Godep is working fine for me now from a clean `GOPATH`, without removing `Godeps/`:

```
GOPATH=/tmp/go
mkdir $GOPATH
godep restore
godep save -v ./... github.com/progrium/go-extpoints
```

vishh commented:
@jfoy: Any updates on this PR? I can help address the godeps issue if you are busy.

jfoy commented:
I would love help addressing the godeps issue. I'm anxious to get this merged.

vishh commented:
@jfoy: godeps can be a pain at times. I went ahead and fixed godeps on HEAD and also included your patches in the process. I will open a separate PR that will include your patches in this PR. Sorry for the trouble with godeps.

piosz commented:
closing this in favor of #687
